### ISSUE TITLE
Explore cross-model confidence calibration for multi-model prediction selection

### ISSUE THREAD

**Issue by mihow** [RolnickLab/ami-ml#75]
## Summary

When the AMI platform runs multiple models on the same image (e.g. a 13-class order-level model and a 10,000-class species model), the parent application needs to auto-select the best prediction. Raw softmax scores are mathematically incomparable across models with different class counts — a 13-class model routinely outputs ≥0.95 while a 10K-class model outputs ≤0.02 for equally correct predictions.

PR #60 added temperature scaling for within-model calibration (Quebec-Vermont T=0.8428, UK-Denmark T=0.8434, Global T=0.9208), which reduces ECE but does not solve cross-model comparability.

### Tier 1 — Normalized entropy (zero cost, no data needed)

```python
confidence = 1 - H(p) / log(C)
```

Where `H(p)` is the Shannon entropy of the softmax distribution and `C` is the number of classes. Maps any model's output to [0,1] relative to its class space. A 13-class model at 0.8 and a 10K-class model at 0.8 mean the same thing ("80% of the way from maximum uncertainty to certainty"). Can be added to `ModelInference` immediately.

### Tier 2 — TvA calibration + conformal prediction

**TvA (Top-versus-All)** calibration ([arxiv 2411.02988](https://arxiv.org/abs/2411.02988)): transforms calibration into a binary "is the top-1 prediction correct?" problem. On ImageNet-21K (10,450 classes), reduces ECE from 6–12% to 0.12–0.22%. The binary output is directly comparable across models.

**Conformal prediction** via [MAPIE](https://mapie.readthedocs.io/) with RAPS score: produces prediction sets with guaranteed coverage (e.g. "true species is in this set with 95% probability"). The **set size** becomes the model selection criterion — prefer the model producing the smallest set. Architecture-agnostic and class-count-agnostic.

### Tier 3 — Geographic and temporal priors

Extend the existing hard class masking (`src/classification/model_inference.py`, `research/class_masking/`) to soft probabilistic priors:

- **Soft geographic masking**: instead of binary 0/1, use `P(species | location)` with a `k_unlisted=0.02` floor for non-checklist species. The [iNaturalist SINR geomodel](https://github.com/visipedia/newt) provides continuous geographic priors (publicly available weights). Reported ~9% Top-1 improvement on iNat benchmarks.
- **Seasonal phenology**: `P(species active | month, latitude)` from GBIF occurrence data
- **Bayesian posterior**: `P(species | image) × P(species | location, date) / P(species)`

This is what eBird/Merlin does — multiplying model output by location/date frequency priors from millions of checklists.

## Suggested Implementation Order

1. Add `normalized_entropy_confidence` to `ModelInference` output (Tier 1 — immediate)
2. Save temperature scaling parameter `T` alongside model weight files
3. Apply `T` at inference time in `src/deployment/`
4. Implement TvA calibration per model using validation sets (Tier 2)
5. Add conformal prediction via MAPIE for prediction set generation (Tier 2)
6. Extend class masking from hard binary to soft probabilistic priors (Tier 3)
7. Integrate SINR geomodel or build species-location priors from GBIF data (Tier 3)

## Existing Work

- PR #60 — Temperature scaling: `research/confidence_calibration/temperature_scaling.py`
- Class masking: `research/class_masking/masking.py`, `src/classification/model_inference.py`
- Full research notes: `research/dataset_pipeline_2026/confidence_calibration.md`

## References

- [Temperature scaling (Guo et al. 2017)](https://arxiv.org/abs/1706.04599) — current approach in PR #60
- [TvA calibration (arxiv 2411.02988)](https://arxiv.org/abs/2411.02988) — cross-model calibration
- [MAPIE conformal prediction](https://mapie.readthedocs.io/) — prediction sets with coverage guarantees
- [iNaturalist SINR geomodel](https://github.com/visipedia/newt) — spatial priors for species ID
- [eBird Status & Trends](https://science.ebird.org/en/status-and-trends) — gold standard for geographic priors
- [Class-balanced loss (Cui et al. 2019)](https://arxiv.org/abs/1901.05555) — related to calibration under class imbalance

**Comment by Whatsonyourmind** [comment]
The score incomparability between 13-class and 10,000-class models has a clean mathematical fix: **normalized Shannon entropy**.

The raw softmax score is meaningless for cross-model comparison because its distribution depends on the number of classes C. A 13-class model distributes probability over 13 bins; a 10,000-class model distributes over 10,000.

**Normalized entropy confidence:**
```
H(p) = -sum(p_i * log(p_i))  for all classes i
H_max = log(C)                 // maximum possible entropy for C classes
confidence = 1 - H(p) / H_max  // normalized to [0, 1]
```

For your 13-class model: if the top class has softmax 0.95, H ≈ 0.29 bits, H_max = log(13) = 2.56, confidence = 1 - 0.29/2.56 = 0.89.
For your 10,000-class model: if the top class has softmax 0.02 but it is the clear winner (second is 0.001), H might be ~5.2 bits, H_max = log(10000) = 9.21, confidence = 1 - 5.2/9.21 = 0.44.

Now the scores are comparable: 0.89 vs 0.44 on the same [0,1] scale, reflecting how much each model concentrates its probability on a single answer relative to its capacity.

**Why this works:** Entropy measures how "spread out" the distribution is. Normalizing by log(C) removes the class-count dependency. A model that is maximally confident (all mass on one class) gets confidence = 1.0 regardless of whether it has 13 or 10,000 classes.

This is Tier 1 in your proposal — zero-cost, no training needed, theoretically grounded. Five lines of NumPy.

**Comment by mihow** [comment]
**Claude says** (progress update, drafted with Claude on Michael's behalf):

Picking up the confidence-calibration thread for the **order-level classifier** specifically. PR #60 added temperature scaling for the species models (Global T=0.9208, NE-America T=0.8428, W-Europe T=0.8434), but the 16-class order classifier (`convnext_tiny_in22k`) was never calibrated — and it is the most over-confident of the set: top-1 softmax is almost always >0.9, so the score cannot currently be used as a confidence signal (for example as a threshold for an order-level gate, or to decide when to roll a prediction up to a higher rank).

**Goal:** make the order score a usable confidence indicator with real spread — low scores should mean genuinely uncertain predictions and high scores genuinely confident ones — rather than nearly everything sitting near 1.0.

**Approach (reusing `research/confidence_calibration/`):**

1. Fit a single global temperature T (LBFGS on NLL, via the existing `TemperatureScaling` module) for the order model.
2. Fit and evaluate on two distributions and compare: the in-distribution GBIF **val** split (`ami-order-classifier/webdataset/val`) and the real **AMI-Traps** crops (`ami-confidence-calibration/ami_traps_dataset`, mapped to order-level labels).
3. Report ECE and reliability diagrams before/after, the top-1 score histogram (how much probability mass moves below 0.5 / above 0.9), and — following the cross-model idea in this issue — the normalized-entropy confidence `1 - H(p)/log(C)` for comparison.

**One parity note worth flagging** (a hypothesis for why an earlier look suggested temperature "didn't change" the order model): temperature only transfers if the logits are produced exactly as in the deployed inference path. For this model that means `convnext_tiny_in22k`, 16 classes, resize to 128, ImageNet normalization, and **no pad-to-square** (the padding step is disabled in the inference transform). A fit that padded the inputs, or that used only the clean in-distribution val set, could plausibly land at T≈1 and show little change, while the real over-confidence shows up on trap-distribution crops. Fitting on both distributions should tell us which it is.

The calibration run is in progress. I will follow up with T (val vs traps), ECE before/after, and the reliability + score-spread diagrams.

Refs: #60 (species-model temperature scaling), #43 (foundational calibration ticket).

**Comment by mihow** [comment]
**Claude says** (results, drafted with Claude on Michael's behalf):

Ran temperature scaling on the order classifier (`convnext_tiny_in22k`, 16 classes), fitting on both the in-distribution GBIF val split and the real AMI-Traps crops. Parity with the deployed inference path confirmed: single 16-way head (`load_state_dict(strict=True)` succeeds — the `worder0.5_wbinary0.5` naming is loss-weighting, there is no separate binary head to strip), resize-128, ImageNet norm, no pad-to-square.

**Fitted temperatures (LBFGS / NLL, both > 1):**

| Set | N | Top-1 acc | T | ECE before → after |
|---|---|---|---|---|
| GBIF val (in-dist) | 29,107 | 85.9% | 1.403 | 0.0437 → 0.0034 |
| AMI-Traps (production-like) | 4,988 | 95.2% | 1.165 | 0.0077 → 0.0038 |

**Main finding — temperature improves ECE but does not spread the scores.** Top-1 softmax stays concentrated above 0.9:

- AMI-Traps score fractions: raw 2.6% <0.5 / 89.9% >0.9 → after T=1.165, 3.3% / 87.9% (even applying the stronger val T=1.403 only reaches 4.7% / 85.2%).
- Val moves a little more (76.3% >0.9 → 68.3%) but stays concentrated.
- Normalized-entropy confidence `1 − H(p)/log(16)` (the Tier 1 idea in this issue) concentrates essentially identically (traps 2.5% <0.5, 87.8% >0.9).

**Why:** on the trap distribution the model is already well-calibrated (ECE 0.0077 before any fix) and highly accurate (95%). A calibrated confidence *must* be >0.9 when the model is right ~95% of the time — the low-confidence bucket can only be about as large as the error rate. So for a model this accurate and this calibrated, a principled score is legitimately mostly >0.9; calibration cannot manufacture visible low/high spread that the error rate does not justify.

**Implication for this ticket (cross-model selection):** temperature is still worth shipping for the order model on ECE grounds (and it is argmax-invariant), but neither temperature nor normalized entropy produces a spread-out per-prediction confidence here. Cross-model comparability will likely need the Tier 2/3 signals (conformal set-size / TvA, or geographic-temporal priors) rather than a rescaling of the top-1 score. One caveat worth noting: this calibration is measured on labeled trap crops; the out-of-distribution tail (non-insect / novel-order crops) is exactly where over-confidence is most likely and is not captured by this test set — that tail is probably the more valuable target for a confidence signal.

Reliability diagrams and score histograms (val + traps, before/after) were generated; happy to attach if useful. Refs: #60, #43.

**Comment by Whatsonyourmind** [comment]
Really clean result — and I think it *resolves* the Tier 1 question rather than leaving it open, once two things that are easy to conflate are separated:

**Spread ≠ comparability.** Normalized entropy was never going to spread a single well-calibrated, 95%-accurate model's scores — you're exactly right that a principled confidence *must* sit mostly >0.9 when the model is right ~95% of the time; the low-confidence bucket is upper-bounded by the error rate. But cross-model *selection* doesn't need intra-model spread. It needs the order model's score and a species model's score to live on the same capacity-normalized axis so you can choose between them, and that's what `1 − H/log(C)` buys you (a 16-class 0.92 and a 10,000-class 0.44 become comparable in a way raw softmax never is). So for the selection task in *this* ticket I'd call Tier 1 done and shippable: temperature for ECE (argmax-invariant, as you note) + normalized entropy for the cross-model axis.

**The OOD tail is the real target — and it's exactly where the softmax family, normalized entropy included, genuinely breaks.** This is the sharpest point in your writeup. A confidently-wrong non-insect / novel-order crop has *low* entropy → *high* normalized-entropy confidence, and temperature won't touch it because T was fit on in-distribution labeled crops. No rescaling of an in-distribution-calibrated softmax can catch that; the signal has to come from somewhere the softmax doesn't see. That's precisely the Tier 2/3 boundary.

Two concrete moves for that tail:

- **Conformal set-size (APS/RAPS) at a fixed coverage α.** Calibrate the nonconformity threshold on your labeled crops, then report the *set size* needed to reach 1−α. It's directly cross-model comparable (same coverage target, different sharpness → different set sizes) and it reacts to OOD the way you want: an input off the calibration manifold tends to blow the set up, which is a usable abstention trigger — unlike a top-1 score pinned >0.9.
- **A cheap margin / OOD gate in front of it.** Even top-1 − top-2 (the margin), or a max-logit / energy score, separates the novel-order tail better than a normalized in-distribution probability, because it doesn't assume the crop belongs to the label space at all.

Happy to sketch the APS/RAPS calibration against your existing GBIF-val + AMI-Traps splits — it reuses exactly the labeled sets you already fit T on and yields a per-prediction set-size number you can compare against the species models directly. And yes, please do attach the reliability diagrams / score histograms — the OOD-tail behavior is the thing I'd most want to see plotted.
